File: js/src/strategies/regex.js
import { normalizePathname, splitSegments, parseQuery } from "../normalize.js";

/**
 * Build a regular expression for a single dynamic route.
 *
 * Each static segment becomes a literal (with regex-special characters
 * escaped); each ":name" segment becomes a capturing group. The pattern is
 * anchored with "^" and "$".
 *
 * @param {{route: string, segments: import("../pattern.js").PatternSegment[]}} route - The compiled dynamic route.
 * @returns {RegExp} The anchored regular expression.
 */
function buildRegex(route) {
  const pattern =
    "^/" +
    route.segments
      .map((segment) => {
        if (segment.type === "static") {
          return segment.value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
        }
        return "([^/]+)";
      })
      .join("/") +
    "$";
  return new RegExp(pattern);
}

/**
 * Create a matcher strategy backed by regular expressions.
 *
 * Suitable for collections with a moderate number of dynamic routes. Static
 * routes use an exact Map lookup, dynamic routes use precompiled regular
 * expressions, and an optional wildcard is used as a fallback.
 *
 * @param {import("../compile.js").CompiledRoutes} compiled - The compiled route collection.
 * @param {boolean} [includeQuery] - When true, parse the query string into the result.
 * @returns {import("../compile.js").MatchStrategy} A strategy exposing a match method.
 */
export function createRegexStrategy(compiled, includeQuery = true) {
  const regexRoutes = compiled.dynamicRoutes.map((route) => ({
    regex: buildRegex(route),
    paramNames: route.paramNames,
    route: route.route,
    value: route.value,
  }));

  return {
    match(pathname) {
      const normalized = normalizePathname(pathname);
      const query = includeQuery ? parseQuery(pathname) : {};

      const staticHit = compiled.staticRoutes.get(normalized);
      if (staticHit) {
        return {
          found: true,
          pathname: normalized,
          route: staticHit.route,
          params: {},
          query,
          value: staticHit.value,
        };
      }

      const segments = splitSegments(normalized);
      const reconstructed = "/" + segments.join("/");

      for (const entry of regexRoutes) {
        const match = entry.regex.exec(reconstructed);
        if (match) {
          const params = {};
          entry.paramNames.forEach((name, index) => {
            params[name] = decodeURIComponent(match[index + 1]);
          });
          return {
            found: true,
            pathname: normalized,
            route: entry.route,
            params,
            query,
            value: entry.value,
          };
        }
      }

      if (compiled.wildcard) {
        return {
          found: true,
          pathname: normalized,
          route: compiled.wildcard.route,
          params: {},
          query,
          value: compiled.wildcard.value,
        };
      }

      return {
        found: false,
        pathname: normalized,
        route: null,
        params: {},
        query,
        value: null,
      };
    },
  };
}

File: python/roution/strategies/regex.py
import re
from urllib.parse import unquote

from roution.normalize import normalize_pathname, split_segments, parse_query
from roution.result import make_result


def _build_regex(dynamic_route):
    parts = []
    for index, segment in enumerate(dynamic_route.segments):
        if index > 0:
            parts.append("/")
        if segment.type == "static":
            parts.append(re.escape(segment.value))
        else:
            parts.append("([^/]+)")
    return re.compile("^/" + "".join(parts) + "$")


class RegexStrategy:
    """Matcher strategy backed by precompiled regular expressions.

    Static routes use an exact dict lookup, dynamic routes use precompiled
    regular expressions, and an optional wildcard is used as a fallback.
    """

    def __init__(self, compiled, include_query):
        self.compiled = compiled
        self.include_query = include_query
        self.regex_routes = [
            (_build_regex(route), route.param_names, route.route, route.value)
            for route in compiled.dynamic_routes
        ]

    def match(self, pathname):
        normalized = normalize_pathname(pathname)
        query = parse_query(pathname) if self.include_query else {}

        static_hit = self.compiled.static_routes.get(normalized)
        if static_hit is not None:
            return make_result(
                True, normalized, static_hit.route, {}, query, static_hit.value
            )

        segments = split_segments(normalized)
        reconstructed = "/" + "/".join(segments)
        for regex, param_names, route, value in self.regex_routes:
            match = regex.match(reconstructed)
            if match is None:
                continue
            params = {
                name: unquote(match.group(index + 1))
                for index, name in enumerate(param_names)
            }
            return make_result(True, normalized, route, params, query, value)

        if self.compiled.wildcard is not None:
            wildcard = self.compiled.wildcard
            return make_result(
                True, normalized, wildcard.route, {}, query, wildcard.value
            )

        return make_result(False, normalized, None, {}, query, None)


def create_regex_strategy(compiled, include_query):
    return RegexStrategy(compiled, include_query)
